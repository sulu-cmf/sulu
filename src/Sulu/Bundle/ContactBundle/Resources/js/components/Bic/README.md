This component accepts nothing but a valid [BIC](https://www.iso9362.org/). It returns `undefined` and will be marked by
a red border if the given string is not a valid BIC. It uses an [`Input` component](#input) underneath, and also exposes
the props from the `Input` making sense for this use case.

```javascript
initialState = {value: undefined};

const onChange = (value) => setState({value});

<div>
    <Bic onChange={onChange} value={state.value} />
    Value: {state.value}
</div>
```
