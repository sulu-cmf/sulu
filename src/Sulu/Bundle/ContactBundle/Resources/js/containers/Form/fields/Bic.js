// @flow
import React from 'react';
import type {FieldTypeProps} from 'sulu-admin-bundle/types';
import BicComponent from '../../../components/Bic';

export default class Bic extends React.Component<FieldTypeProps<?string>> {
    render() {
        const {dataPath, disabled, error, onChange, onFinish, value} = this.props;

        return (
            <BicComponent
                disabled={!!disabled}
                id={dataPath}
                onBlur={onFinish}
                onChange={onChange}
                valid={!error}
                value={value}
            />
        );
    }
}
