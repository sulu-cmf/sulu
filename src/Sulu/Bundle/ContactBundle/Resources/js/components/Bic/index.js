// @flow
import Bic from './Bic';

export default Bic;
