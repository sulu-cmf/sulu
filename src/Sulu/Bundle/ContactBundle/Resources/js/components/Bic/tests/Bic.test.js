// @flow
import React from 'react';
import {render, mount} from 'enzyme';
import Bic from '../Bic';

test('Bic should render', () => {
    const onChange = jest.fn();
    expect(render(<Bic onChange={onChange} value={null} />)).toMatchSnapshot();
});

test('Bic should render with placeholder', () => {
    expect(render(<Bic onChange={jest.fn()} placeholder="My placeholder" value={null} />)).toMatchSnapshot();
});

test('Bic should render with value', () => {
    const bic = mount(<Bic onChange={jest.fn()} value="BBBBCCLLXXX" />);
    expect(bic.render()).toMatchSnapshot();
});

test('Bic should render when disabled', () => {
    expect(render(<Bic disabled={true} onChange={jest.fn()} value="BBBBCCLLXXX" />)).toMatchSnapshot();
});

test('Bic should render error', () => {
    expect(render(<Bic onChange={jest.fn()} valid={false} value={null} />)).toMatchSnapshot();
});

test('Bic should render error when invalid value is set', () => {
    const bic = mount(<Bic onChange={jest.fn()} value={null} />);

    // check if showError is set correctly
    bic.find('Input').instance().props.onChange('xxx', {target: {value: 'xxx'}});
    bic.find('Input').instance().props.onBlur();
    bic.update();
    expect(bic.instance().showError).toBe(true);

    // now add a valid value
    bic.find('Input').instance().props.onChange('BBBBCCLLXXX', {target: {value: 'BBBBCCLLXXX'}});
    bic.find('Input').instance().props.onBlur();
    bic.update();
    expect(bic.instance().showError).toBe(false);
});

test('Bic should trigger callbacks correctly', () => {
    const onChange = jest.fn();
    const onBlur = jest.fn();
    const bic = mount(<Bic onBlur={onBlur} onChange={onChange} value={null} />);

    // provide invalid value
    bic.find('Input').instance().props.onChange('xxx', {target: {value: 'xxx'}});
    bic.find('Input').instance().props.onBlur();
    bic.update();
    expect(onChange).toHaveBeenLastCalledWith(undefined);
    expect(onBlur).toBeCalled();

    // provide one more invalid value
    bic.find('Input').instance().props.onChange('BBBBCCLLX', {target: {value: 'BBBCCLLX'}});
    bic.find('Input').instance().props.onBlur();
    bic.update();
    expect(onChange).toHaveBeenLastCalledWith(undefined);
    expect(onBlur).toBeCalled();

    // now add a valid value
    bic.find('Input').instance().props.onChange('BBBBCCLLXXX', {target: {value: 'BBBBCCLLXXX'}});
    bic.find('Input').instance().props.onBlur();
    bic.update();
    expect(onChange).toHaveBeenLastCalledWith('BBBBCCLLXXX');
    expect(onBlur).toBeCalled();

    // provide one more valid value
    bic.find('Input').instance().props.onChange('BBBBCCLL', {target: {value: 'BBBBCCLL'}});
    bic.find('Input').instance().props.onBlur();
    bic.update();
    expect(onChange).toHaveBeenLastCalledWith('BBBBCCLL');
    expect(onBlur).toBeCalled();

    expect(onBlur).toHaveBeenCalledTimes(4);
});
