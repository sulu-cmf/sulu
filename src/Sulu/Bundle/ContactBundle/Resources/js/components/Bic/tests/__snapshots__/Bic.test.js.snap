// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bic should render 1`] = `
<label
  class="input default left"
>
  <div
    class="prependedContainer default"
  >
    <span
      aria-label="su-earth"
      class="su-earth icon default"
    />
  </div>
  <input
    type="text"
    value=""
  />
</label>
`;

exports[`Bic should render error 1`] = `
<label
  class="input default left error"
>
  <div
    class="prependedContainer default"
  >
    <span
      aria-label="su-earth"
      class="su-earth icon default"
    />
  </div>
  <input
    type="text"
    value=""
  />
</label>
`;

exports[`Bic should render when disabled 1`] = `
<label
  class="input default left disabled"
>
  <div
    class="prependedContainer default"
  >
    <span
      aria-label="su-earth"
      class="su-earth icon default"
    />
  </div>
  <input
    disabled=""
    type="text"
    value=""
  />
</label>
`;

exports[`Bic should render with placeholder 1`] = `
<label
  class="input default left"
>
  <div
    class="prependedContainer default"
  >
    <span
      aria-label="su-earth"
      class="su-earth icon default"
    />
  </div>
  <input
    placeholder="My placeholder"
    type="text"
    value=""
  />
</label>
`;

exports[`Bic should render with value 1`] = `
<label
  class="input default left"
>
  <div
    class="prependedContainer default"
  >
    <span
      aria-label="su-earth"
      class="su-earth icon default"
    />
  </div>
  <input
    type="text"
    value="BBBBCCLLXXX"
  />
</label>
`;
